heartbeat/monitor/monitor.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/ElrondNetwork/elrond-go/common"
+	"github.com/ElrondNetwork/elrond-go/heartbeat"
+	"github.com/ElrondNetwork/elrond-go/heartbeat/data"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+var log = logger.GetOrCreate("heartbeat/monitor")
+
+const minDuration = time.Second
+
+// ArgHeartbeatV2Monitor holds the arguments needed to create a new instance of heartbeatV2Monitor
+type ArgHeartbeatV2Monitor struct {
+	Cache                         storage.Cacher
+	PubKeyConverter               core.PubkeyConverter
+	Marshaller                    marshal.Marshalizer
+	PeerShardMapper               process.PeerShardMapper
+	MaxDurationPeerUnresponsive   time.Duration
+	HideInactiveValidatorInterval time.Duration
+	ShardId                       uint32
+}
+
+type heartbeatV2Monitor struct {
+	cache                         storage.Cacher
+	pubKeyConverter               core.PubkeyConverter
+	marshaller                    marshal.Marshalizer
+	peerShardMapper               process.PeerShardMapper
+	maxDurationPeerUnresponsive   time.Duration
+	hideInactiveValidatorInterval time.Duration
+	shardId                       uint32
+	numInstances                  map[string]uint64
+}
+
+// NewHeartbeatV2Monitor creates a new instance of heartbeatV2Monitor
+func NewHeartbeatV2Monitor(args ArgHeartbeatV2Monitor) (*heartbeatV2Monitor, error) {
+	err := checkArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	return &heartbeatV2Monitor{
+		cache:                         args.Cache,
+		pubKeyConverter:               args.PubKeyConverter,
+		marshaller:                    args.Marshaller,
+		peerShardMapper:               args.PeerShardMapper,
+		maxDurationPeerUnresponsive:   args.MaxDurationPeerUnresponsive,
+		hideInactiveValidatorInterval: args.HideInactiveValidatorInterval,
+		shardId:                       args.ShardId,
+		numInstances:                  make(map[string]uint64, 0),
+	}, nil
+}
+
+func checkArgs(args ArgHeartbeatV2Monitor) error {
+	if check.IfNil(args.Cache) {
+		return heartbeat.ErrNilCacher
+	}
+	if check.IfNil(args.PubKeyConverter) {
+		return heartbeat.ErrNilPubkeyConverter
+	}
+	if check.IfNil(args.Marshaller) {
+		return heartbeat.ErrNilMarshaller
+	}
+	if check.IfNil(args.PeerShardMapper) {
+		return heartbeat.ErrNilPeerShardMapper
+	}
+	if args.MaxDurationPeerUnresponsive < minDuration {
+		return fmt.Errorf("%w on MaxDurationPeerUnresponsive, provided %d, min expected %d",
+			heartbeat.ErrInvalidTimeDuration, args.MaxDurationPeerUnresponsive, minDuration)
+	}
+	if args.HideInactiveValidatorInterval < minDuration {
+		return fmt.Errorf("%w on HideInactiveValidatorInterval, provided %d, min expected %d",
+			heartbeat.ErrInvalidTimeDuration, args.HideInactiveValidatorInterval, minDuration)
+	}
+
+	return nil
+}
+
+// GetHeartbeats returns the heartbeat status
+func (monitor *heartbeatV2Monitor) GetHeartbeats() []data.PubKeyHeartbeat {
```

Comment:
this is not a concurrent safe function. Because the usage of the internal monitor.numInstances map this struct might panic the node, if 2 or more requests are done in the same time.
Either remove this property map, or add mutex protection here

Reply:
:+1:  removed the property map as it was not needed anyway